javalib/src/main/scala/java/util/Formatter.scala

```diff
@@ -79,1587 +91,989 @@ class Formatter(
+  }
 
-      if (null == arg)
-        return transformFromNull()
+  @inline private def validateFlagsForPercentAndNewline(
+      flags: Flags,
+      conversion: Char,
+      invalidFlags: Int): Unit = {
 
-      val calendar =
-        arg match {
-          case cal: Calendar => cal
-          case _ =>
-            val date =
-              arg.asInstanceOf[Any] match {
-                case l: Long => new Date(l)
-                case d: Date => d
-                case _ =>
-                  throw new IllegalFormatConversionException(
-                    currentConversionType,
-                    arg.getClass())
-              }
-            val calendar = Calendar.getInstance(locale)
-            calendar.setTime(date)
-            calendar
-        }
+    @noinline def illegalFlags(): Nothing =
+      throw new IllegalFormatFlagsException(flagsToString(flags))
 
-      if (null == dateTimeUtil)
-        dateTimeUtil = new DateTimeUtil(locale)
-      val result = new StringBuilder()
-      dateTimeUtil.transform(formatToken, calendar, result)
-      padding(result, startIndex)
-    }
+    if ((flags.bits & invalidFlags) != 0)
+      illegalFlags()
   }
 
-  private object Transformer {
-    private var lineSeparator: String = _
+  /* Should in theory be a method of `Flags`. See the comment on that class
+   * about why we keep it here.
+   */
+  private def flagsToString(flags: Flags): String = {
+    (if (flags.leftAlign) "-" else "") +
+      (if (flags.altFormat) "#" else "") +
+      (if (flags.positivePlus) "+" else "") +
+      (if (flags.positiveSpace) " " else "") +
+      (if (flags.zeroPad) "0" else "") +
+      (if (flags.useGroupingSeps) "," else "") +
+      (if (flags.negativeParen) "(" else "") +
+      (if (flags.useLastIndex) "<" else "")
   }
 
-  private class FloatUtil(private var result: StringBuilder,
-                          private var formatToken: FormatToken,
-                          private val decimalFormat: DecimalFormat,
-                          private val argument: Object) {
-    private var minusSign =
-      decimalFormat.getDecimalFormatSymbols().getMinusSign()
-
-    def transform(aFormatToken: FormatToken, aResult: StringBuilder): Unit = {
-      this.result = aResult
-      this.formatToken = aFormatToken
-      formatToken.getConversionType() match {
-        case 'e' | 'E' => transform_e()
-        case 'f'       => transform_f()
-        case 'g' | 'G' => transform_g()
-        case 'a' | 'A' => transform_a()
-        case _ =>
-          throw new UnknownFormatConversionException(
-            String.valueOf(formatToken.getConversionType()))
-      }
+  private def computerizedScientificNotation(
+      x: Double,
+      precision: Int,
+      forceDecimalSep: Boolean): String = {
+    val s1: String = Zone { implicit z =>
+      val buf = z.alloc(2 + precision)
+      vsprintf(buf, toCString(s"%.${precision}e"), toCVarArgList(x))
+      fromCString(buf)
+    }
```

Comment:
For this, we might have to preserve *some* of the logic that was in `java.text.DecimalFormat`. The previous code was *way* too complicated, because it built a pattern string, which was then re-parsed by `DecimalFormat.applyPattern`. Plus, `DecimalFormat.format` wants to deal with localization itself, but here we should hard-code for ROOT (`.` for decimal sep, no grouping, ASCII digits); the localization is handled later by `LocaleLocaleInfo.localizeNumber`.

So here we should be able to retain only a very minimal portion of what used to be `DecimalFormat`.